File: Python/MinMaxStack.py
class MinMaxStack:
    def __init__(self):
        self.minmaxstack = []
        self.stack = []

    def peek(self):
        return self.stack[len(self.stack)-1]

    def pop(self):
        self.minmaxstack.pop()
        return self.stack.pop()

    def push(self, num):
        newMinMax = {'min': num, 'max': num}
        if len(self.minmaxstack):
            lastminmax = self.minmaxstack[len(self.minmaxstack)-1]
            newMinMax['min'] = min(lastminmax['min'], num)
            newMinMax['max'] = max(lastminmax['max'], num)
        self.minmaxstack.append(newMinMax)
        self.stack.append(num)

    def getMin(self):
        return self.minmaxstack[len(self.minmaxstack)-1]['min']

    def getMax(self):
        return self.minmaxstack[len(self.minmaxstack)-1]['max']


if __name__ == "__main__":
    x = MinMaxStack()
    x.push(11)
    x.push(3)
    x.push(5)
    print(x.getMin())
    print(x.getMax())
    print(x.pop())

File: MinMaxStack.java
//First lets create our stack node

import java.util.ArrayList;
import java.util.List;

class StNode {
    int val;
    int prevMin;
    int prevMax;

    public StNode() {
    }

    public StNode(int val, int prevMin, int prevMax) {
        this.val = val;
        this.prevMin = prevMin;
        this.prevMax = prevMax;
    }
}

class MinMaxStack {

    List<StNode> st;
    int min = Integer.MAX_VALUE;
    int max = Integer.MIN_VALUE;

    public MinMaxStack() {
        this.st = new ArrayList<>();
    }

    public void push(int val) {
        // Lets push the val
        StNode newOne = new StNode();
        newOne.val = val;
        newOne.prevMin = min;
        newOne.prevMax = max;
        min = Math.min(min, val);
        max = Math.max(max, val);
        st.add(newOne);
    }

    public void pop() {
        // get the last one,remove and set new min and max
        StNode lastOne = st.remove(st.size() - 1);
        max = lastOne.prevMax;
        min = lastOne.prevMin;
    }

    public int top() {
        return st.get(st.size() - 1).val;
    }

    public int getMin() {
        return min;
    }

    public int getMax() {
        return max;
    }

    public static void main(String[] args) {
        MinMaxStack obj = new MinMaxStack();
        obj.push(2);
        obj.push(0);
        obj.push(-1);
        obj.pop();
        System.out.println(obj.getMin());
        System.out.println(obj.getMax());
    }
}
